### PR TITLE
Tweak MSC implementation

### PR DESCRIPTION
**Pull request by sethrj**
I should have caught or commented on this in review, but I spent more time trying to understand the MSC implementation as I was working on the "action ID" refactor.
- Move the MscStep/MscInteraction  results into the Interaction.hh for now -- better than `detail` since they're being used well outside the "detail" scope
- Build the MSC test (but disable running it) when vecgeom is unavailable
- Save MSC step to global memory, then restore after linear propagation, rather than keeping as local variable
- Be more careful to separate "non-random" MSC calculations from ones that require sampling, as well as separating input data from temporary state data

Perhaps after validating the new MSC code we should refactor it:
- put the path length transformation helpers into one class and add tests
- likewise make all the functions that have conditionals and min/max more testable

We're definitely not testing enough for the number of conditions present: we should add some thin unit test interfaces and plot results as a function of the inputs (such as energy) for the many piecewise functions.

### CODE REVIEW

File: src/physics/em/detail/UrbanMscData.hh [hunk 1]
```diff
@@ -34,7 +34,7 @@ struct UrbanMscParameters
     real_type range_fact{0.04}; //!< range_factor for e-/e+ (0.2 for muon/h)
     real_type safety_fact{0.6}; //!< safety factor
     real_type safety_tol{0.01}; //!< safety tolerance
-    real_type geom_limit{5e-8 * units::millimeter}; //!< 0.05 nm
+    real_type geom_limit{5e-8 * units::millimeter}; //!< minimum boundary
```
Comment (whokion) [inline src/physics/em/detail/UrbanMscData.hh:37]: Is "minimum boundary" right choice?  May be "minimum msc step".

File: src/physics/em/detail/UrbanMscScatter.hh [hunk 2]
```diff
@@ -76,13 +61,19 @@ class UrbanMscScatter
...
+    real_type lambda_;
+    real_type true_path_;
+
+    // Internal state
+    real_type tau_{0};
+
     //// COMMON PROPERTIES ////
 
-    //! The minimum step length for geometry ( 0.05*CLHEP::nm)
+    //! The minimum step length for geometry 0.05 nm
```
Comment (whokion) [inline src/physics/em/detail/UrbanMscScatter.hh:76]: I find that this should be same as UrbanMscParameters::geom_limit (which I mistakenly introduced it redundantly).  Can you choose one of them and remove the other?

**Review by whokion — APPROVED** [review]
@sethrj  Nice improvement!  Please note one critical comment for converting geom_path to true path after LinearPropagator.

**Comment by sethrj** [comment]
@whokion The last changes could use another look if you would please. I think the MSC code needs a lot more unit testing and assertions: there are so many `min`,`max`,`if` in there, and so many embedded distributions being sampled...

File: src/physics/em/detail/UrbanMscScatter.hh [hunk 3]
```diff
@@ -163,48 +177,43 @@ UrbanMscScatter::UrbanMscScatter(const UrbanMscRef&       shared,
...
     {
-        auto end_energy = helper_.end_energy(true_path_);
+        // Do not sample scattering at the last or at a small step
+        return {true_path_, inc_direction_, {0, 0, 0}};
+    }
 
-        bool skip_sampling = (end_energy.value() < 1e-9
-                              || true_path_ <= input_.limit_min
-                              || true_path_ < lambda_ * params_.tau_small);
+    // Ssample azimuthal angle, used for displacement and exiting angle
```
Comment (whokion) [inline src/physics/em/detail/UrbanMscScatter.hh:186]: Ssample ==> Sample

File: src/physics/em/detail/UrbanMscScatter.hh [hunk 4]
```diff
@@ -163,48 +177,43 @@ UrbanMscScatter::UrbanMscScatter(const UrbanMscRef&       shared,
...
 {
-    MscInteraction result{true_path_, inc_direction_, {0, 0, 0}};
-
-    // Do not sample scattering at the last or at a small step
-    if (true_path_ < helper_.range() && true_path_ > geom_min())
+    if (skip_sampling_)
     {
-        auto end_energy = helper_.end_energy(true_path_);
+        // Do not sample scattering at the last or at a small step
+        return {true_path_, inc_direction_, {0, 0, 0}};
```
Comment (whokion) [inline src/physics/em/detail/UrbanMscScatter.hh:183]: return MscInteraction{true_path_, inc_direction_, {0, 0, 0}};  (to be consistent with the return at the bottom?).

File: src/physics/em/detail/UrbanMscScatter.hh [hunk 5]
```diff
@@ -163,48 +177,43 @@ UrbanMscScatter::UrbanMscScatter(const UrbanMscRef&       shared,
...
         }
     }
+    else
+    {
+        displacement = {0, 0, 0};
+    }
 
-    return result;
+    // Sample direction
+    real_type mu = this->sample_cos_theta(
```
Comment (whokion) [inline src/physics/em/detail/UrbanMscScatter.hh:211]: Is mu better than cth? cos_theta? Either way is okay.

File: test/physics/em/UrbanMsc.test.cc [hunk 6]
```diff
@@ -285,24 +285,22 @@ TEST_F(UrbanMscTest, msc_scattering)
...
-    const double expected_angle[] = {0.000314741326035635,
-                                     0.738624667826603,
-                                     -0.145610123961716,
-                                     -0.657415795200799,
-                                     0.119014628205848,
-                                     -0.216102964881347,
-                                     0.793645871128744,
-                                     -0.98020130119347};
-
+    static const double expected_fstep[] = {0.0027916899999997,
```
Comment (whokion) [inline test/physics/em/UrbanMsc.test.cc:288]: Now results are quite different, which infers that tweaking codes in this MR introduced something logically different from original codes (which I cannot identify easily now, but we need to make sure that something is not wrongly changed - unless the random sequence is changed).

**Review by whokion — COMMENTED** [review]
@sethrj I do agree that we need a detail validation as there are (too) many different call paths for different input conditions - I kind of try to test with different step lengths and energies but do agree that the current test is minimal and far from covering many cases - the test geometry may be too simple too.  I will definitely do more verification as required.

Besides different results (observed in UrbanMscTest) from this MR, demo-loop with msc_enabled also crashes with an error message, "..../celeritas/src/base/Collection.hh:363:
celeritas: internal assertion failed: i < this->size()" (it looks that https://github.com/sethrj/celeritas/blob/msc-tweaks/app/demo-loop/LDemoLauncher.hh#L163  is problematic even though the backtrace hints there is a problem TrackInitData.hh line 179).   Also,  
` msc_step = other.msc_step`
should be added after
https://github.com/sethrj/celeritas/blob/msc-tweaks/src/physics/base/PhysicsData.hh#L348

File: src/physics/em/detail/UrbanMscScatter.hh (hunk 5 above)
Comment (sethrj) [reply src/physics/em/detail/UrbanMscScatter.hh:211]: `$\mu = cos \theta$` is the canonical way of representing polar angle cosines in radiation transport in nuclear engineering... but I guess it's not common in HEP (I don't see it in Geant4). I can revert this usage, but note that there was one function declaration written as `phi` but was called with `cth`.

File: test/physics/em/UrbanMsc.test.cc (hunk 6 above)
Comment (sethrj) [reply test/physics/em/UrbanMsc.test.cc:288]: I made a note in the final commit, but this is strictly due to the ordering of the RNG sampling (which I think uses fewer temporaries now). Moving the polar angle sampling *before* the azimuthal angle restores the original results.

**Comment by sethrj** [comment]
Thanks for catching the crash: it's because I added new state data but didn't add new code for resizing (or copying) it.

**Comment by sethrj** [comment]
OK, I made a temporary build with debugging and enabled MSC in the input, and it works. I'd forgotten how little testing we have on the transporter... and the switch with MSC means we have no testing. I'm going to make it a priority to:
- move the Transporter to the main codebase 
- add a primitive safety distance calculation for ORANGE (for geometry that has only "simple" volumes, only planes/spheres/cylinders) 
- make an easier way to do "regression" tests

File: src/physics/em/detail/UrbanMscScatter.hh (hunk 5 above)
Comment (whokion) [reply src/physics/em/detail/UrbanMscScatter.hh]: Okay, then let's keep mu - always better to know new/various conventions.

File: test/physics/em/UrbanMsc.test.cc (hunk 6 above)
Comment (whokion) [reply test/physics/em/UrbanMsc.test.cc:288]: Make a sense if the sampling order has changed.

**Review by whokion — APPROVED** [review]
@sethrj Looks good to me and confirm that domo-loop with msc runs with no problem.  Thanks for many improvements!

**Pull request merged**